fastcrypto-zkp/src/bn254/zk_login.rs

```diff
@@ -0,0 +1,753 @@
+/// The field value is computed from the max key claim length and its provider.
+static SUPPORTED_KEY_CLAIM_TO_FIELD: Lazy<HashMap<(&str, String), &str>> = Lazy::new(|| {
+    let mut map = HashMap::new();
+    map.insert(
+        (
+            OAuthProvider::Google.get_config().0,
+            SupportedKeyClaim::Sub.to_string(),
+        ),
+        "18523124550523841778801820019979000409432455608728354507022210389496924497355",
+    );
+    map.insert(
+        (
+            OAuthProvider::Google.get_config().0,
+            SupportedKeyClaim::Email.to_string(),
+        ),
+        "",
+    );
+    map.insert(
+        (
+            OAuthProvider::Twitch.get_config().0,
+            SupportedKeyClaim::Sub.to_string(),
+        ),
+        "",
+    );
+    map.insert(
+        (
+            OAuthProvider::Twitch.get_config().0,
+            SupportedKeyClaim::Email.to_string(),
+        ),
+        "",
+    );
+    map
+});
+
+/// Supported OAuth providers. Must contain "openid" in "scopes_supported"
+/// and "public" for "subject_types_supported" instead of "pairwise".
+#[derive(Debug)]
+pub enum OAuthProvider {
+    Google, // https://accounts.google.com/.well-known/openid-configuration
+    Twitch, // https://id.twitch.tv/oauth2/.well-known/openid-configuration
+}
+
+impl OAuthProvider {
+    /// Returns a tuple of iss string and JWK endpoint string for the given provider.
+    pub fn get_config(&self) -> (&str, &str) {
+        match self {
+            OAuthProvider::Google => (
+                "https://accounts.google.com",
+                "https://www.googleapis.com/oauth2/v2/certs",
+            ),
+            OAuthProvider::Twitch => (
+                "https://id.twitch.tv/oauth2",
+                "https://id.twitch.tv/oauth2/keys",
+            ),
+        }
+    }
+
+    pub fn from_iss(iss: &str) -> Result<Self, FastCryptoError> {
+        match iss {
+            "https://accounts.google.com" => Ok(Self::Google),
+            "https://id.twitch.tv/oauth2" => Ok(Self::Twitch),
+            _ => Err(FastCryptoError::InvalidInput),
+        }
+    }
+}
+
+/// The claims in the body signed by OAuth provider that must
+/// be locally unique to the provider and cannot be reassigned.
+#[derive(Debug)]
+pub enum SupportedKeyClaim {
+    Sub,
+    Email,
+}
+
+impl fmt::Display for SupportedKeyClaim {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            SupportedKeyClaim::Email => write!(f, "email"),
+            SupportedKeyClaim::Sub => write!(f, "sub"),
+        }
+    }
+}
+
+pub fn get_supported_claims() -> Vec<String> {
+    let supported_claims: Vec<SupportedKeyClaim> =
+        vec![SupportedKeyClaim::Sub, SupportedKeyClaim::Email];
+
+    supported_claims
+        .iter()
+        .map(|claim| claim.to_string())
+        .collect()
+}
+
+/// Verify proof using fixed verifying key.
+pub fn verify_groth16_with_fixed_vk(
```

Comment:
Do we plan to invoke this from Sui? I'm currently ok to have special functions and fixed configs for the time being here, maybe in the future we should put them in a different "gadgets" folder in fastcrypto or even in Sui. Good to proceed as it is atm.